### PR TITLE
[mysql] Snapshot split the chunks asynchronously (ververica#915)

### PR DESCRIPTION
**Pull request by fuyun2024**

### CODE REVIEW

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java [hunk 1]
```diff
@@ -180,6 +189,28 @@ private void captureNewlyAddedTables() {
...
+            if (executor == null) {
+                ThreadFactory threadFactory =
+                        new ThreadFactoryBuilder().setNameFormat("snapshot-split").build();
+                this.executor = Executors.newSingleThreadExecutor(threadFactory);
+            }
+
+            executor.submit(
+                    () -> {
+                        while (remainingTables.size() > 0) {
+                            TableId nextTable = remainingTables.pollFirst();
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java:203]: How we make sure the table splits generation has finished if failover happened ?

Comment (fuyun2024) [reply flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java]: Thank you for your comments. I have revised the code again.

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java [hunk 2]
```diff
@@ -180,27 +191,60 @@ private void captureNewlyAddedTables() {
...
+            synchronized (lock) {
+                try {
+                    lock.wait();
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
             }
+            return getNext();
+        } else {
+            return Optional.empty();
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java:241]: Could we release the executor resource after splitting finished ?

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java [hunk 3]
```diff
@@ -180,27 +191,60 @@ private void captureNewlyAddedTables() {
...
-            } else {
-                return Optional.empty();
+        } else if (!remainingTables.isEmpty()) {
+            // wait for the asynchronous split to complete
+            synchronized (lock) {
+                try {
+                    lock.wait();
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java:237]: please throw `FlinkRuntimeException` with a detail error message.

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java [hunk 4]
```diff
@@ -180,27 +191,60 @@ private void captureNewlyAddedTables() {
...
-            TableId nextTable = remainingTables.pollFirst();
-            if (nextTable != null) {
-                // split the given table into chunks (snapshot splits)
-                Collection<MySqlSnapshotSplit> splits = chunkSplitter.generateSplits(nextTable);
-                remainingSplits.addAll(splits);
-                alreadyProcessedTables.add(nextTable);
-                return getNext();
-            } else {
-                return Optional.empty();
+        } else if (!remainingTables.isEmpty()) {
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java:230]: There's a race condition that `remainingTables` is empty but the asynchronous thread just added the snapshot splits of last table to `remainingSplits`, and then the last table will be missed.

**Review by leonardBang — COMMENTED** [review]
Thanks @fuyun2024 for the contribution, could you also add some unit tests to cover the change?

**Review by leonardBang — APPROVED** [review]
Thanks @fuyun2024 for the contribution, LGTM

**Pull request merged**